### PR TITLE
Range Query

### LINKED ISSUE #11: Add support for range queries
> name:[Jones TO Smith]
> inclusive range; selects documents whose name field has any value between Jones and Smith, including boundaries.
> 
> score:{2.5 TO 7.3}
> exclusive range; selects documents whose score field is between 2.5 and 7.3, excluding boundaries.
> 
> score:{2.5 TO *]
> one-sided range; selects documents whose score field is larger than 2.5. 

https://lucene.apache.org/core/9_4_2/queryparser/org/apache/lucene/queryparser/flexible/standard/StandardQueryParser.html

These examples are actually already in [StandardQueryParserDocsSuite](https://github.com/cozydev-pink/lucille/blob/main/core/src/test/scala/io/pig/lucille/StandardQueryParserDocsSuite.scala#L185-L198) as failing tests.

### PR DESCRIPTION
**Pull request by samspills**
- adds `RangeQ` class
- adds `rangeQuery` parser
  - parses inclusive or exclusive boundaries
  - parses wildcard boundary value (`*`)
  - parses boundary values as strings `(a-zA-Z0-9.)`

resolves #11 
<details><summary>old notes</summary>

~no where close to working yet, just parsing a single string~
pretty close to working 😎 

### Need Help With
- [x] how do I connect to the rest of the query parsing? If I pass in `query: Parser[Query]` instead of a string, how do I make that connect and work? 

### Sam's TODOs (Andrew don't look)
- [x] currently lower and upper are parsing into a non empty list of characters, want to parse into either `String` or `Float` or `Int` I guess?
  - does this make sense actually?
  - if lucene allows `{5 TO Janeway}` then maybe parsing as more than a string doesn't make sense? Need more testing I think
  - chatted with Andrew, just parse as a string for now
- [x] currently not handling a possible `*` for a bound
- [x] is a lowercase `to` allowed? currently doesn't parse
  - I think this is correct, the lucene query parser only documents `TO`, and boolean operators _must_ be all caps. Should probably test against lucene to be sure, but for now this is fine
- [x] is there a way to parse the spaces and maybe spaces, but drop all of the units out of the map? That is a WILD looking tuple yo

</details>

### CODE REVIEW

File: core/src/main/scala/pink/cozydev/lucille/Parser.scala [hunk 1]
```diff
@@ -115,6 +115,23 @@ object Parser {
   def unaryMinus(query: P[Query]): P[UnaryMinus] =
     P.char('-') *> query.map(UnaryMinus.apply)
 
+  // Range query
+  // e.g. '{cats TO dogs}', '[1 TO *}'
```
Comment (valencik) [inline core/src/main/scala/pink/cozydev/lucille/Parser.scala:119]: Ooohh, thanks for including this :)

File: core/src/main/scala/pink/cozydev/lucille/Parser.scala [hunk 2]
```diff
@@ -115,6 +115,23 @@ object Parser {
...
+  def rangeQuery: P[RangeQ] = {
+    val inclLower = P.charIn('{', '[').map(lowerBound => lowerBound == '[') <* maybeSpace
+    val inclUpper = maybeSpace *> P.charIn('}', ']').map(upperBound => upperBound == ']')
+    val boundValue =
+      P.char('*').as(None) | P.not(P.stringIn(reserved)).with1 *> (alpha | digit | P.char(
```
Comment (valencik) [inline core/src/main/scala/pink/cozydev/lucille/Parser.scala:124]: We had some discussion about whether or not we should allow `{cats TO NOT}`, and decided to leave this as is for now and check with Lucene.
I kind of suspect it's allowed as a negation doesn't make sense in that position. But we should check to be sure.

**Review by valencik — APPROVED** [review]
This is awesome, I'm excited to use this. Thanks a bunch!
